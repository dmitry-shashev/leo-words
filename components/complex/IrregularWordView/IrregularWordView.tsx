'use client'

import { FC } from 'react'
import { useSelector } from 'react-redux'
import {
  getCurrentIrregularWord,
  getIrregularWordStep,
} from '@/store/slices/irregularWordsSlice'
import { IrregularWordStep } from '@/types/irregular-word-step'

function determineWordSizeClass(word: string): string {
  const length = word.length
  if (length > 16) {
    return 'text-xl'
  }
  if (length > 24) {
    return 'text-lg'
  }
  return 'text-3xl'
}

export const IrregularWordView: FC = () => {
  const currentWord = useSelector(getCurrentIrregularWord)
  const irregularWordStep = useSelector(getIrregularWordStep)

  return (
    <div className="mb-10">
      <div
        className={`${determineWordSizeClass(currentWord.infinitive)} text-center h-40 text-blue-400`}
      >
        <div aria-label="Infinitive word" key={currentWord.infinitive}>
          {currentWord.infinitive}
        </div>
      </div>

      <div
        className={`${determineWordSizeClass(currentWord.pastSimple)} text-center h-40 text-blue-400`}
      >
        {[
          IrregularWordStep.CHECK_PAST_SIMPLE.value,
          IrregularWordStep.CHECK_PAST_PARTICIPLE.value,
        ].includes(irregularWordStep.value) && (
          <div aria-label="Past simple word" key={currentWord.pastSimple}>
            {currentWord.pastSimple}
          </div>
        )}
      </div>

      <div
        className={`${determineWordSizeClass(currentWord.pastParticiple)} text-center h-44 text-blue-400`}
      >
        {[IrregularWordStep.CHECK_PAST_PARTICIPLE.value].includes(
          irregularWordStep.value
        ) && (
          <div
            aria-label="Past participle word"
            key={currentWord.pastParticiple}
          >
            {currentWord.pastParticiple}
          </div>
        )}
      </div>
    </div>
  )
}
