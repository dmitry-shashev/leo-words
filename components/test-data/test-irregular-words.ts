import { IrregularWord } from '@/models/irregular-word'

export const testIrregularWords: Array<IrregularWord> = [
  {
    id: 1,
    infinitive: 'bite',
    pastSimple: 'bit',
    pastParticiple: 'bitten',
  },
  {
    id: 2,
    infinitive: 'draw',
    pastSimple: 'drew',
    pastParticiple: 'drawn',
  },
  {
    id: 3,
    infinitive: 'do',
    pastSimple: 'did',
    pastParticiple: 'done',
  },
  {
    id: 4,
    infinitive: 'fly',
    pastSimple: 'flew',
    pastParticiple: 'flown',
  },
  {
    id: 5,
    infinitive: 'drive',
    pastSimple: 'drove',
    pastParticiple: 'driven',
  },
]
