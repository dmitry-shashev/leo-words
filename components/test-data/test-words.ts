import { Word } from '@/models/word'

export const testWords: Array<Word> = [
  {
    id: 31684,
    combinedTranslation: 'проповедовать',
    wordValue: 'preach',
    transcription: 'prˈiːtʃ',
    picture: 'https://contentcdn.lingualeo.com/uploads/picture/634277.png',
    pronunciation: 'https://audiocdn.lingualeo.com/v2/3/cHJlYWNo6ba93e74.mp3',
    created: 1433462400,
    added: 1433462400,
  },
  {
    id: 35298,
    combinedTranslation: 'польза',
    wordValue: 'sake',
    transcription: 'seɪk',
    picture: 'https://contentcdn.lingualeo.com/uploads/picture/326972.png',
    pronunciation: 'https://audiocdn.lingualeo.com/v2/3/c2FrZQ==30ec034a.mp3',
    created: 1433462400,
    added: 1433462400,
  },
  {
    id: 39328,
    combinedTranslation: 'прогулка',
    wordValue: 'stroll',
    transcription: 'stroʊl',
    picture: 'https://contentcdn.lingualeo.com/uploads/picture/2898240.png',
    pronunciation: 'https://audiocdn.lingualeo.com/v2/3/c3Ryb2xsf9e1435d.mp3',
    created: 1433462400,
    added: 1433462400,
  },
  {
    id: 43022,
    combinedTranslation: 'предпринимать',
    wordValue: 'undertake',
    transcription: 'ˈʌndərˌteɪk',
    picture: 'https://contentcdn.lingualeo.com/uploads/picture/19282.png',
    pronunciation:
      'https://audiocdn.lingualeo.com/v2/3/dW5kZXJ0YWtlce0a68dd.mp3',
    created: 1433462400,
    added: 1433462400,
  },
  {
    id: 54653,
    combinedTranslation: 'эффективно',
    wordValue: 'effectively',
    transcription: 'ɪˈfɛktɪvli',
    picture: 'https://contentcdn.lingualeo.com/uploads/picture/31663.png',
    pronunciation:
      'https://audiocdn.lingualeo.com/v2/3/ZWZmZWN0aXZlbHk=c9b5216a.mp3',
    created: 1433462400,
    added: 1433462400,
  },
]
