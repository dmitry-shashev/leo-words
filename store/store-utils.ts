import { Word } from '@/models/word'
import { IrregularWord } from '@/models/irregular-word'

export function getEmptyWord(): Word {
  return {
    id: 0,
    wordValue: '',
    pronunciation: '',
    picture: '',
    combinedTranslation: '',
    transcription: '',
    created: 0,
  }
}

export function getEmptyIrregularWord(): IrregularWord {
  return {
    id: 0,
    infinitive: '',
    pastSimple: '',
    pastParticiple: '',
  }
}
