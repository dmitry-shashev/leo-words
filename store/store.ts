import {
  combineReducers,
  configureStore,
  EnhancedStore,
} from '@reduxjs/toolkit'
import settingsReducer from './slices/settingsSlice'
import wordsReducer from './slices/wordsSlice'
import irregularWordsReducer from './slices/irregularWordsSlice'
import { useDispatch } from 'react-redux'
import storage from 'redux-persist/lib/storage'
import {
  FLUSH,
  REHYDRATE,
  PAUSE,
  PERSIST,
  PURGE,
  REGISTER,
  persistStore,
  persistReducer,
} from 'redux-persist'

const persistConfig = {
  key: 'root',
  storage,
}

const reducer = combineReducers({
  settingsReducer,
  wordsReducer,
  irregularWordsReducer,
})

const persistedReducer = persistReducer(persistConfig, reducer)

export function createStore(): EnhancedStore {
  return configureStore({
    reducer: persistedReducer,
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware({
        serializableCheck: {
          ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
        },
      }),
  })
}

const store = createStore()
let persistor = persistStore(store)

export type RootState = ReturnType<typeof store.getState>
export type AppDispatch = typeof store.dispatch
export const useAppDispatch: () => AppDispatch = useDispatch

export { store, persistor }
