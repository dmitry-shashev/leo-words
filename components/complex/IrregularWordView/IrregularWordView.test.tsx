import { test, expect } from 'vitest'
import { createStore } from '@/store/store'
import { act, render } from '@testing-library/react'
import { Provider } from 'react-redux'
import { IrregularWordView } from '@/components/complex/IrregularWordView/IrregularWordView'
import {
  goToNextIrregularWordStep,
  setAllIrregularWords,
} from '@/store/slices/irregularWordsSlice'
import { testIrregularWords } from '@/components/test-data/test-irregular-words'

test('IrregularWordView', () => {
  const store = createStore()
  store.dispatch(setAllIrregularWords(testIrregularWords))

  const { asFragment } = render(
    <Provider store={store}>
      <IrregularWordView />
    </Provider>
  )

  expect(asFragment()).toMatchSnapshot()

  act(() => {
    store.dispatch(goToNextIrregularWordStep())
  })
  expect(asFragment()).toMatchSnapshot()

  act(() => {
    store.dispatch(goToNextIrregularWordStep())
  })
  expect(asFragment()).toMatchSnapshot()
})
