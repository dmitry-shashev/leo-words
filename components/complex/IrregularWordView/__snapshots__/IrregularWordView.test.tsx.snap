// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`IrregularWordView 1`] = `
<DocumentFragment>
  <div
    class="mb-10"
  >
    <div
      class="text-3xl text-center h-40 text-blue-400"
    >
      <div
        aria-label="Infinitive word"
      >
        bite
      </div>
    </div>
    <div
      class="text-3xl text-center h-40 text-blue-400"
    />
    <div
      class="text-3xl text-center h-44 text-blue-400"
    />
  </div>
</DocumentFragment>
`;

exports[`IrregularWordView 2`] = `
<DocumentFragment>
  <div
    class="mb-10"
  >
    <div
      class="text-3xl text-center h-40 text-blue-400"
    >
      <div
        aria-label="Infinitive word"
      >
        bite
      </div>
    </div>
    <div
      class="text-3xl text-center h-40 text-blue-400"
    >
      <div
        aria-label="Past simple word"
      >
        bit
      </div>
    </div>
    <div
      class="text-3xl text-center h-44 text-blue-400"
    />
  </div>
</DocumentFragment>
`;

exports[`IrregularWordView 3`] = `
<DocumentFragment>
  <div
    class="mb-10"
  >
    <div
      class="text-3xl text-center h-40 text-blue-400"
    >
      <div
        aria-label="Infinitive word"
      >
        bite
      </div>
    </div>
    <div
      class="text-3xl text-center h-40 text-blue-400"
    >
      <div
        aria-label="Past simple word"
      >
        bit
      </div>
    </div>
    <div
      class="text-3xl text-center h-44 text-blue-400"
    >
      <div
        aria-label="Past participle word"
      >
        bitten
      </div>
    </div>
  </div>
</DocumentFragment>
`;
